examples/defaulter-gen/_output_tests/marker/zz_generated.go

```diff
@@ -0,0 +1,95 @@
+// +build !ignore_autogenerated
+
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Code generated by defaulter-gen. DO NOT EDIT.
+
+package marker
+
+import (
+	"encoding/json"
+	"reflect"
+
+	runtime "k8s.io/apimachinery/pkg/runtime"
+)
+
+// RegisterDefaults adds defaulters functions to the given scheme.
+// Public to allow building arbitrary schemes.
+// All generated defaulters are covering - they call all nested defaulters.
+func RegisterDefaults(scheme *runtime.Scheme) error {
+	scheme.AddTypeDefaultingFunc(&Defaulted{}, func(obj interface{}) { SetObjectDefaults_Defaulted(obj.(*Defaulted)) })
+	scheme.AddTypeDefaultingFunc(&DefaultedWithFunction{}, func(obj interface{}) { SetObjectDefaults_DefaultedWithFunction(obj.(*DefaultedWithFunction)) })
+	return nil
+}
+
+func SetObjectDefaults_Defaulted(in *Defaulted) {
+	if reflect.ValueOf(in.Field).IsZero() {
+		in.Field = "bar"
+	}
+	if reflect.ValueOf(in.OtherField).IsZero() {
```

Comment:
Hm, these ValueOf calls will probably be slow / make garbage. Can you instead pre-compute the zero value and use == to tell?

Alternatively, you could make a benchmark that says I'm wrong :)

Reply:
What about this:
```golang
package test

import (
	"reflect"
	"testing"
)

func Benchmark_Valueof(b *testing.B) {
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if reflect.ValueOf(n + 1).IsZero() {
			b.Fatalf("%v is not 0", n+1)
		}
	}
}

func Benchmark_direct(b *testing.B) {
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if n+1 == 0 {
			b.Fatalf("%v is not 0", n+1)
		}
	}
}
```
Gives:
```shell
$ go test -bench=. ./...
goos: darwin
goarch: amd64
Benchmark_Valueof-8   	72421857	        15.5 ns/op	       8 B/op	       0 allocs/op
Benchmark_direct-8    	1000000000	         0.252 ns/op	       0 B/op	       0 allocs/op
```

Reply:
hm, if it's really 0 bytes, then maybe this isn't a problem, can you try this with something more complicated like a map, struct, slice... ?

I don't think I care about the CPU time, although that's like 60x slower and these functions call this for every defaulted field (?). 66 fields and we're up to a millisecond, that's a real amount of time.

@wojtek-t @deads2k any thoughts on what our perf budget should be for default setting?

Reply:
In the immediate term, we're only planning on applying this setting to containerPort.protocol, since this is the default that is causing us trouble. As we extend and start using this, we can start optimizing as needed.

> can you try this with something more complicated like a map, struct, slice...

Sure.

Reply:
OK if we're only using it for one field in 1.20, that should be fine.

Reply:
Thanks for the benchmarks Antoine! I'll add this as an action item for future improvements/optimizations.